libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt

```diff
@@ -0,0 +1,139 @@
+package com.mapbox.navigation.core.reroute
+
+import androidx.annotation.MainThread
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.base.common.logger.Logger
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
+import com.mapbox.navigation.core.directions.session.DirectionsSession
+import com.mapbox.navigation.core.directions.session.RoutesRequestCallback
+import com.mapbox.navigation.core.routeoptions.RouteOptionsProvider
+import com.mapbox.navigation.core.trip.session.TripSession
+import com.mapbox.navigation.utils.internal.JobControl
+import com.mapbox.navigation.utils.internal.ThreadController
+import java.util.concurrent.CopyOnWriteArraySet
+import kotlinx.coroutines.launch
+
+/**
+ * Default implementation of [RerouteController]
+ */
+internal class MapboxRerouteController(
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession,
+    private val routeOptionsProvider: RouteOptionsProvider,
+    threadController: ThreadController = ThreadController,
+    private val logger: Logger
+) : RerouteController {
+
+    private val observers = CopyOnWriteArraySet<RerouteController.RerouteStateObserver>()
+
+    private val mainJobController: JobControl = threadController.getMainScopeAndRootJob()
+
+    override var state: RerouteState = RerouteState.Idle
+        private set(value) {
+            if (field == value) {
+                return
+            }
+            field = value
+            observers.forEach { it.onRerouteStateChanged(field) }
+        }
+
+    private companion object {
+        const val TAG = "MapboxRerouteController"
+    }
+
+    // current implementation ignore `onNewRoutes` callback because `DirectionsSession` update routes internally
+    override fun reroute(routesCallback: RerouteController.RoutesCallback) {
+        state = RerouteState.FetchingRoute
+        logger.d(
+            Tag(TAG),
+            Message("Fetching route")
+        )
+        routeOptionsProvider.update(
+            directionsSession.getRouteOptions(),
+            tripSession.getRouteProgress(),
+            tripSession.getEnhancedLocation()
+        )
+            .let { routeOptionsResult ->
+                when (routeOptionsResult) {
+                    is RouteOptionsProvider.RouteOptionsResult.Success -> {
+                        request(routeOptionsResult.routeOptions)
+                    }
+                    is RouteOptionsProvider.RouteOptionsResult.Error -> {
+                        mainJobController.scope.launch {
```

Comment:
I think we spread all events for user in UI thread I believe it's a contract for our SDK(discussed it long time ago), right @LukasPaczos ?

Reply:
Yes, we should default to the main thread unless there's a strong reason not to.

Reply:
Can we move it to https://github.com/mapbox/mapbox-navigation-android/blob/dee64b14cf3473da11a9f7977734b7dc2713128d/libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt#L33-L40 so we don't need to call `mainJobController.scope.launch` on every place we're setting the `state`? Also could you cut a ticket to keep track of 👇?

> Asking because we're not doing that for the rest of the observers that can be registered through `MapboxNavigation` 🤔

Reply:
in this case state will be changed from different thread, and we get situation when `state = X`, but observers have received `state = Y`, moreover, in this case  we need
```
@Volatile
var state: RerouteState
```

Reply:
> Also could you cut a ticket to keep track of 👇?
>
>> Asking because we're not doing that for the rest of the observers that can be registered through `MapboxNavigation` 🤔

We should investigate if that's needed in other parts of the codebase.

Reply:
I believe it's already handing on UI thread, but we should be ok to make audit 👍 https://github.com/mapbox/mapbox-navigation-android/issues/3294